src/priceClient/adapters/defiLlama.ts

```diff
@@ -0,0 +1,92 @@
+import assert from "assert";
+import { PriceFeedAdapter, TokenPrice } from "../priceClient";
+import { BaseHTTPAdapter } from "./baseAdapter";
+
+type DefiLlamaTokenPrice = {
+  price: number;
+  symbol: string;
+  timestamp: number;
+  confidence: number;
+};
+
+type DefiLlamaPriceResponse = {
+  coins: {
+    [key: string]: DefiLlamaTokenPrice;
+  };
+};
+
+type DefiLlamaArgs = {
+  name?: string;
+  host?: string;
+  timeout?: number;
+  minConfidence?: number;
+};
+
+export class PriceFeed extends BaseHTTPAdapter implements PriceFeedAdapter {
+  private _minConfidence: number;
+
+  constructor({
+    name = "DefiLlama",
+    host = "coins.llama.fi",
+    timeout = 5000,
+    minConfidence = 0.9,
+  }: DefiLlamaArgs = {}) {
+    super(name, host, { timeout });
+    assert(minConfidence >= 0.0 && minConfidence <= 1.0);
+    this._minConfidence = minConfidence;
+  }
+
+  get minConfidence(): number {
+    return this._minConfidence;
+  }
+
+  set minConfidence(minConfidence: number) {
+    assert(minConfidence >= 0.0 && minConfidence <= 1.0);
+    this._minConfidence = minConfidence;
+  }
+
+  async getPriceByAddress(address: string, currency = "usd"): Promise<TokenPrice> {
+    const price = await this.getPricesByAddress([address], currency);
+    return price[0];
+  }
+
+  async getPricesByAddress(addresses: string[], currency = "usd"): Promise<TokenPrice[]> {
+    if (currency != "usd") throw new Error(`Currency ${currency} not supported by ${this.name}`);
+
+    const path = "prices/current/" + addresses.map((address) => `ethereum:${address}`).join();
+    const tokenPrices: unknown = await this.query(path, {});
+    if (!this.validateResponse(tokenPrices))
+      throw new Error(`Unexpected ${this.name} response: ${JSON.stringify(tokenPrices)}`);
+
+    return addresses.map((address: string) => {
+      const tokenPrice = tokenPrices.coins[`ethereum:${address}`];
```

Comment:
Is the request casing -> response casing relationship documented? I think this is fine as is since we'll start seeing errors if the address requested is missing.

An alternative would be to convert all keys to lowercase before processing them.

Reply:
> Is the request casing -> response casing relationship documented? I think this is fine as is since we'll start seeing errors if the address requested is missing.

Third parties don't really document this - it's only been identified via probing their responses, which of course means it's potentially subject to change.

It's not explicitly documented anywhere in the adapters either, but the PriceClient notes that it stores all addresses in lower-case internally: https://github.com/across-protocol/sdk-v2/blob/77269818898a691623e1a567ed55dae3011be9ae/src/priceClient/priceClient.ts#L18

Because the PriceClient generates its list of addresses from its internal cache, this means that the adapters will implicitly always receive lower-case (...per current implementation).

https://github.com/across-protocol/sdk-v2/blob/77269818898a691623e1a567ed55dae3011be9ae/src/priceClient/priceClient.ts#L78

 I'll concede that it's not really obvious, though...and I was scratching my head a bit on exactly this question earlier. I'll think about how I can describe it properly.

Reply:
Has been documented here: https://github.com/across-protocol/sdk-v2/pull/128/commits/12a81b81bec2e4d7f6a1115df5a87a77213629ff.